drivers/bluetooth/hci/hci_sf32lb.c

```diff
@@ -0,0 +1,613 @@
+{
+	struct bt_sf32lb_data *hci = dev->data;
+
+	hci_read_hdr(dev);
+
+	if (!hci->rx.remaining) {
+		struct bt_hci_sco_hdr *hdr = &hci->rx.sco;
+
+		hci->rx.remaining = hdr->len;
+		LOG_DBG("Got SCO header. Payload %u bytes", hci->rx.remaining);
+		hci->rx.have_hdr = true;
+	}
+}
+
+static inline void get_iso_hdr(const struct device *dev)
+{
+	struct bt_sf32lb_data *hci = dev->data;
+
+	hci_read_hdr(dev);
+
+	if (!hci->rx.remaining) {
+		struct bt_hci_iso_hdr *hdr = &hci->rx.iso;
+
+		hci->rx.remaining = bt_iso_hdr_len(sys_le16_to_cpu(hdr->len));
+		LOG_DBG("Got ISO header. Payload %u bytes", hci->rx.remaining);
+
+		hci->rx.have_hdr = true;
+	}
+}
+
+static inline void get_evt_hdr(const struct device *dev)
+{
+	struct bt_sf32lb_data *hci = dev->data;
+
+	struct bt_hci_evt_hdr *hdr = &hci->rx.evt;
+
+	hci_read_hdr(dev);
+
+	if (hci->rx.hdr_len == sizeof(*hdr) && hci->rx.remaining < sizeof(*hdr)) {
+		switch (hci->rx.evt.evt) {
+		case BT_HCI_EVT_LE_META_EVENT:
+			hci->rx.remaining++;
+			hci->rx.hdr_len++;
+			break;
+#if defined(CONFIG_BT_CLASSIC)
+		case BT_HCI_EVT_INQUIRY_RESULT_WITH_RSSI:
+		case BT_HCI_EVT_EXTENDED_INQUIRY_RESULT:
+			hci->rx.discardable = true;
+			break;
+#endif
+		default:
+			break;
+		}
+	}
+
+	if (!hci->rx.remaining) {
+		if (hci->rx.evt.evt == BT_HCI_EVT_LE_META_EVENT &&
+		    ((hci->rx.hdr[sizeof(*hdr)] == BT_HCI_EVT_LE_ADVERTISING_REPORT) ||
+		     (hci->rx.hdr[sizeof(*hdr)] == BT_HCI_EVT_LE_EXT_ADVERTISING_REPORT))) {
+			LOG_DBG("Marking adv report as discardable");
+			hci->rx.discardable = true;
+		}
+
+		hci->rx.remaining = hdr->len - (hci->rx.hdr_len - sizeof(*hdr));
+		LOG_DBG("Got event header. Payload %u bytes", hdr->len);
+		hci->rx.have_hdr = true;
+	}
+}
+
+static inline void copy_hdr(struct bt_sf32lb_data *hci)
+{
+	net_buf_add_mem(hci->rx.buf, hci->rx.hdr, hci->rx.hdr_len);
+}
+
+static void reset_rx(struct bt_sf32lb_data *hci)
+{
+	hci->rx.type = BT_HCI_H4_NONE;
+	hci->rx.remaining = 0U;
+	hci->rx.have_hdr = false;
+	hci->rx.hdr_len = 0U;
+	hci->rx.discardable = false;
+}
+
+static struct net_buf *get_rx(struct bt_sf32lb_data *hci, k_timeout_t timeout)
+{
+	LOG_DBG("type 0x%02x, evt 0x%02x", hci->rx.type, hci->rx.evt.evt);
+
+	switch (hci->rx.type) {
+	case BT_HCI_H4_EVT:
+		return bt_buf_get_evt(hci->rx.evt.evt, hci->rx.discardable, timeout);
+	case BT_HCI_H4_ACL:
+		return bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	case BT_HCI_H4_SCO:
+		if (IS_ENABLED(CONFIG_BT_CLASSIC)) {
+			LOG_ERR("SCO not supported by host stack.");
+		}
+		break;
+	case BT_HCI_H4_ISO:
+		if (IS_ENABLED(CONFIG_BT_ISO)) {
+			return bt_buf_get_rx(BT_BUF_ISO_IN, timeout);
+		}
+		break;
+	default:
+		LOG_ERR("Invalid rx type 0x%02x", hci->rx.type);
+	}
+
+	return NULL;
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	const struct device *dev = p1;
+	struct bt_sf32lb_data *data = dev->data;
+
+	zbt_config_mailbox(dev);
+	lcpu_power_on();
+	k_sleep(K_MSEC(50));
```

Comment:
This looks quite hacky. Where does this need to sleep come from? Is it in some specification? Why specifically 50ms?